File: apps/admin/src/lib/order-status.ts
import type { KnownOrderStatus, OrderStatus } from "@/lib/api/types";
import { ORDER_STATUSES, isKnownOrderStatus } from "@/lib/api/types";

/**
 * Human labels for all 17 statuses in the contract's status vocabulary table,
 * copied verbatim from that table.
 */
export const ORDER_STATUS_LABELS: Record<KnownOrderStatus, string> = {
  submitted: "Submitted",
  under_review: "Under review",
  approved: "Approved",
  inventory_allocated: "Inventory allocated",
  picking: "Picking",
  partially_fulfilled: "Partially fulfilled",
  picked: "Picked",
  packed: "Packed",
  route_assigned: "Route assigned",
  out_for_delivery: "Out for delivery",
  delivered: "Delivered",
  receiving_required: "Receiving required",
  completed: "Completed",
  backordered: "Backordered",
  cancelled: "Cancelled",
  rejected: "Rejected",
  delivery_failed: "Delivery failed",
};

/**
 * Fall back to a readable rendering of an unrecognised status instead of
 * crashing or showing a raw enum value — the contract requires clients to
 * degrade gracefully here.
 */
export function orderStatusLabel(status: OrderStatus): string {
  if (isKnownOrderStatus(status)) return ORDER_STATUS_LABELS[status];
  if (!status) return "Unknown";
  return status
    .replace(/[_-]+/g, " ")
    .replace(/^./, (character) => character.toUpperCase());
}

/** Options for the status filter control, in workflow order. */
export const ORDER_STATUS_OPTIONS: { value: KnownOrderStatus; label: string }[] =
  ORDER_STATUSES.map((status) => ({
    value: status,
    label: ORDER_STATUS_LABELS[status],
  }));

/**
 * Statuses from which HQ may still approve or reject. Anything further along
 * the workflow (or already terminal) is not actionable from this dashboard.
 */
export const ACTIONABLE_ORDER_STATUSES = [
  "submitted",
  "under_review",
  "approved",
] as const satisfies readonly KnownOrderStatus[];

export function isActionableOrderStatus(status: OrderStatus): boolean {
  return (ACTIONABLE_ORDER_STATUSES as readonly string[]).includes(status);
}

File: apps/mobile/lib/shared/models/order_status.dart
/// The full order status vocabulary from the API contract (17 values).
///
/// The wire value is a plain string; models keep the raw string so an
/// unrecognised status from a newer API can still be displayed rather than
/// crashing deserialization. Use [OrderStatus.tryParse] to resolve one.
enum OrderStatus {
  submitted('submitted', 'Submitted'),
  underReview('under_review', 'Under review'),
  approved('approved', 'Approved'),
  inventoryAllocated('inventory_allocated', 'Inventory allocated'),
  picking('picking', 'Picking'),
  partiallyFulfilled('partially_fulfilled', 'Partially fulfilled'),
  picked('picked', 'Picked'),
  packed('packed', 'Packed'),
  routeAssigned('route_assigned', 'Route assigned'),
  outForDelivery('out_for_delivery', 'Out for delivery'),
  delivered('delivered', 'Delivered'),
  receivingRequired('receiving_required', 'Receiving required'),
  completed('completed', 'Completed'),
  backordered('backordered', 'Backordered'),
  cancelled('cancelled', 'Cancelled'),
  rejected('rejected', 'Rejected'),
  deliveryFailed('delivery_failed', 'Delivery failed');

  const OrderStatus(this.wireValue, this.label);

  /// The exact string the API sends and expects.
  final String wireValue;

  /// Human readable label, exactly as tabulated in the contract.
  final String label;

  /// Returns the matching status, or `null` for an unrecognised value.
  static OrderStatus? tryParse(String? value) {
    if (value == null) return null;
    for (final status in OrderStatus.values) {
      if (status.wireValue == value) return status;
    }
    return null;
  }

  /// A best-effort label for any wire value, including unknown ones.
  ///
  /// `some_new_status` degrades to `Some new status` rather than throwing.
  static String labelFor(String? value) {
    final known = tryParse(value);
    if (known != null) return known.label;
    if (value == null || value.isEmpty) return 'Unknown';
    final words = value.replaceAll('_', ' ').trim();
    if (words.isEmpty) return 'Unknown';
    return words[0].toUpperCase() + words.substring(1);
  }

  /// Terminal states: no further progress is expected.
  bool get isTerminal => switch (this) {
    OrderStatus.completed ||
    OrderStatus.cancelled ||
    OrderStatus.rejected ||
    OrderStatus.deliveryFailed => true,
    _ => false,
  };
}
